### PR TITLE
A complementary intersection propagator for filtering the sets.

### PR DESCRIPTION
**Pull request by JLiangWaterloo**
This PR is with regards to issue #448.

I know in your message you said to specialize it for the binary case, but I need it for the n-ary case.

I've ran about a million fuzz tests for the intersection constraint. Based on these tests, I think PropIntersection and PropIntersectionFilterSets are both idempotent. In addition, I think the combination of these two propagators are bound consistent. I couldn't find a counterexample in the million fuzz tests.

### CODE REVIEW

**Comment by cprudhom** [comment]
Thank you for the new propagator.
In order to merge the PR, we also appreciate if you : 
- could add it to the constraint declaration signature in `ISetConstraintFactory`,
- and could integrate tests to the test suite ?

File: src/main/java/org/chocosolver/solver/constraints/set/PropIntersectionFilterSets.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+            sdm[idxVarInProp].forEach(j -> {
+                if (!intersection.getUB().contains(j)) {
+                    SetVar uniqueSet = findUniqueSetThatDoesNotContainJInLB(j);
+                    if (uniqueSet != null) {
+                        uniqueSet.remove(j, this);
+                    }
+                }
+            }, SetEventType.ADD_TO_KER);
+        } else {
+            sdm[idxVarInProp].forEach(j -> {
```
Comment (jgFages) [inline src/main/java/org/chocosolver/solver/constraints/set/PropIntersectionFilterSets.java]: Does this create a new lambda at each propagation? Doesn't it have an overhead compared to creating an object once and for all in the constructor?

**Comment by jgFages** [comment]
Thank you also about your remark on the idempotency of PropIntersection. I have to admit that I do not remember what made us think it was not idempotent. Maybe a bug which has been fixed since then. I believe you can replace the redundant second propIntersection propagator with your new propagator. However, if you remove the second propIntersection propagator, I would feel more confortable if the tests you will add to the suite also cover the idempotency question in addition to the correctness check. Thanks again

**Comment by JLiangWaterloo** [comment]
Done.

I'm not sure how you feel about black box fuzz tests in your testing suite.

**Comment by JLiangWaterloo** [comment]
I just wanted to add one comment: the fuzz tests I typically run for my project are more "random" than the ones in this CL for two reasons:

1. I use a random SetValueSelector for my own fuzz tests, which the Choco library does not have.
2. I have a special implementation of SetStrategy that randomizes whether to enforce-first or remove-first per decision. The current implementation of SetStrategy cannot randomly switch between enforce-first and remove-first since it must stick with the choice passed through its constructor.

**Pull request merged**